apps/hash-graph/bin/cli/built-in-types.jsonl

```diff
@@ -0,0 +1,8 @@
+{"type":"snapshot","blockProtocolModuleVersions":{"graph":"0.3.0"}}
+{"type":"dataType","schema":{"$schema":"https://blockprotocol.org/types/modules/graph/0.3/schema/data-type","kind":"dataType","$id":"https://blockprotocol.org/@blockprotocol/types/data-type/boolean/v/1","title":"Boolean","description":"A True or False value","type":"boolean"},"metadata":{"recordId":{"baseUrl":"https://blockprotocol.org/@blockprotocol/types/data-type/boolean/","version":1},"custom":{"provenance":{"recordCreatedById":"00000000-0000-0000-0000-000000000000"},"temporalVersioning":{"transactionTime":{"start":{"kind":"inclusive","limit":"2023-06-02T17:20:56.452456000Z"},"end":{"kind":"unbounded"}}},"fetchedAt":"2023-06-02T17:20:56.450025000Z"}}}
+{"type":"dataType","schema":{"$schema":"https://blockprotocol.org/types/modules/graph/0.3/schema/data-type","kind":"dataType","$id":"https://blockprotocol.org/@blockprotocol/types/data-type/empty-list/v/1","title":"Empty List","description":"An Empty List","type":"array","const":[]},"metadata":{"recordId":{"baseUrl":"https://blockprotocol.org/@blockprotocol/types/data-type/empty-list/","version":1},"custom":{"provenance":{"recordCreatedById":"00000000-0000-0000-0000-000000000000"},"temporalVersioning":{"transactionTime":{"start":{"kind":"inclusive","limit":"2023-06-02T17:20:57.251893000Z"},"end":{"kind":"unbounded"}}},"fetchedAt":"2023-06-02T17:20:57.249456000Z"}}}
+{"type":"dataType","schema":{"$schema":"https://blockprotocol.org/types/modules/graph/0.3/schema/data-type","kind":"dataType","$id":"https://blockprotocol.org/@blockprotocol/types/data-type/null/v/1","title":"Null","description":"A placeholder value representing 'nothing'","type":"null"},"metadata":{"recordId":{"baseUrl":"https://blockprotocol.org/@blockprotocol/types/data-type/null/","version":1},"custom":{"provenance":{"recordCreatedById":"00000000-0000-0000-0000-000000000000"},"temporalVersioning":{"transactionTime":{"start":{"kind":"inclusive","limit":"2023-06-02T17:20:56.733422000Z"},"end":{"kind":"unbounded"}}},"fetchedAt":"2023-06-02T17:20:56.730679000Z"}}}
+{"type":"dataType","schema":{"$schema":"https://blockprotocol.org/types/modules/graph/0.3/schema/data-type","kind":"dataType","$id":"https://blockprotocol.org/@blockprotocol/types/data-type/number/v/1","title":"Number","description":"An arithmetical value (in the Real number system)","type":"number"},"metadata":{"recordId":{"baseUrl":"https://blockprotocol.org/@blockprotocol/types/data-type/number/","version":1},"custom":{"provenance":{"recordCreatedById":"00000000-0000-0000-0000-000000000000"},"temporalVersioning":{"transactionTime":{"start":{"kind":"inclusive","limit":"2023-06-02T17:20:56.196069000Z"},"end":{"kind":"unbounded"}}},"fetchedAt":"2023-06-02T17:20:56.193720000Z"}}}
```

Comment:
Actually, I think the `fetchedAt` should precisely the time when they were fetched. If you look at the database compare them to the ontology types hosted at blockprotocol.org and see some inconsistency (for whatever reason) it's known that at the `fetchedAt` date the schema looked differently. Why it would cause strange behavior? That's the date of retrieval, not the date of when the types were uploaded to blockprotocol.org.

Reply:
Hm this is a complicated topic, and one that depends on the semantics of `fetchedAt` and the temporal axes. Let's say I start my production instance of my application today, and I start to heavily utilise temporal versioning, creating dashboards over provenance, etc. If I _**know**_ my application's history starts today, then having records that were somehow 'fetched' before my application started is pretty... strange.

I hear your point about inconsistencies.. I mean, according to the spec the types _**must**_ be immutable, but there is always a possibility, and if this ends up used for more than the primitive types it becomes more problematic. I'd almost prefer if there wasn't a `fetchedAt` at all if they are built-in, but that'd introduce a tonne of unnecessary application complexity regarding the external ontology type shape, so let's not do that.

Reply:
But if you would have started the application prior this date, the built-in types should be in the database already.

Reply:
Yes but I mean the first time you start up the app, not restarting it during a migration or crash or whatever :) anyway I'm not sure we have any easy avenues in front of us for how we would mitigate this even if we wanted to, so perhaps it doesn't matter

Reply:
Sorry, but I think we're talking past each other. Why would the behavior be different when rerunning the application?

Reply:
The behaviour is not different, I thought you were getting confused about that, because of your comment here:
> But if you would have started the application prior this date, the built-in types should be in the database already.

Which seemed pretty unrelated to what I was saying 😅 

> Let's say I start my production instance of my application today, and I start to heavily utilise temporal versioning, creating dashboards over provenance, etc. If I know my application's history starts today, [...]

Would it make more sense to you if I add this?

> Let's say I start my production instance of my application today _for the first time ever_ [...]

I don't think this is purely philosophical by the way, although it is very much an edge-case. When dealing with timestamps and such it's important to know the limit points of your data, if any. That's part of why we ran into problems with the time crate, because it was trying to account for years before `0` (i.e. years BCE), which is meaningless in the use-cases we'd encounter or support in our app. Disagreements on the start of time is problematic, because you can set various invariants in your code, and it affects your ability to reason about things in the datastore.

If I launch my application today, therefore starting its history today, it could be problematic if I have things in my database that claim to have entered the database even before it even existed. What if I write migration logic or something to rollback elements that happened after timestamp X, etc.